chunk/schema.go

```diff
@@ -0,0 +1,429 @@
+	result := []ReadBucket{}
+	s.buckets(s, from, through, userID, metricName, func(tableName, hashKey string, startTime model.Time) {
+		result = append(result, s.newReadBucket(tableName, hashKey, startTime, labelName))
+	})
+	return result
+}
+
+func hourlyBuckets(s originalSchema, from, through model.Time, userID string, metricName model.LabelValue, callback func(tableName, hashKey string, startTime model.Time)) {
+	var (
+		fromHour    = from.Unix() / secondsInHour
+		throughHour = through.Unix() / secondsInHour
+	)
+
+	for i := fromHour; i < throughHour; i++ {
+		callback(
+			s.tableForBucket(i*secondsInHour),
+			fmt.Sprintf("%s:%d:%s", userID, i, metricName),
+			model.TimeFromUnix(i*secondsInHour),
+		)
+	}
+}
+
+func dailyBuckets(s originalSchema, from, through model.Time, userID string, metricName model.LabelValue, callback func(tableName, hashKey string, startTime model.Time)) {
+	var (
+		fromDay    = from.Unix() / secondsInDay
+		throughDay = through.Unix() / secondsInDay
+	)
+
+	for i := fromDay; i < throughDay; i++ {
+		callback(
+			s.tableForBucket(i*secondsInDay),
+			fmt.Sprintf("%s:d%d:%s", userID, i, metricName),
+			model.TimeFromUnix(i*secondsInDay),
+		)
+	}
+}
+
+func (s originalSchema) tableForBucket(bucketStart int64) string {
+	if !s.cfg.UsePeriodicTables || bucketStart < (s.cfg.PeriodicTableStartAt.Unix()) {
+		return s.cfg.OriginalTableName
+	}
+	// TODO remove reference to time package here
+	return s.cfg.TablePrefix + strconv.Itoa(int(bucketStart/int64(s.cfg.TablePeriod/time.Second)))
+}
+
+type originalBucket struct {
+	tableName string
+	hashKey   string
+	startTime model.Time
+	labelName model.LabelName
+}
+
+func newOriginalWriteBucket(tableName, hashKey string, startTime model.Time) WriteBucket {
+	return &originalBucket{
+		tableName: tableName,
+		hashKey:   hashKey,
+		startTime: startTime,
+	}
+}
+
+func newOriginalReadBucket(tableName, hashKey string, startTime model.Time, labelName model.LabelName) ReadBucket {
+	return &originalBucket{
+		tableName: tableName,
+		hashKey:   hashKey,
+		startTime: startTime,
+		labelName: labelName,
+	}
+}
+
+func (b *originalBucket) TableName() string     { return b.tableName }
+func (b *originalBucket) HashKey() string       { return b.hashKey }
+func (b *originalBucket) StartTime() model.Time { return b.startTime }
+
+func (b *originalBucket) AllLabelsPrefix() ([]byte, error) {
+	if b.labelName != "" {
+		panic("AllLabelsPrefix called on bucket bound to specific label!")
+	}
+	return nil, nil
+}
+
+func (b *originalBucket) AllValuesPrefix() ([]byte, error) {
+	if b.labelName == "" {
+		panic("AllValuePrefix called on bucket not bound to specific label!")
+	}
+	return buildRangeKey(string(b.labelName)), nil
+}
+
+func (b *originalBucket) SpecificValuePrefix(value model.LabelValue) ([]byte, error) {
+	if b.labelName == "" {
+		panic("SpecificValuePrefix called on bucket not bound to specific label!")
+	}
```

Comment:
I don't have a constructive alternative (yet), but these panics point to the interface not being quite right.